### PR TITLE
Refactor CSV reader benchmarks with nvbench

### LINKED ISSUE #10941: Separate cuIO IO benchmarks from column type benchmarks
Currently cuIO benchmarks cover a cartesian product of IO options and column types, which leads to a large number of cases and long execution time. We can separate the IO options from the column type to decrease the total number of cases.

- [ ] CSV reader #11678 
- [x] ORC reader #11543 
- [x] ORC writer #11598
- [x] Parquet reader #11611 
- [ ] Parquet writer #11623

### PR DESCRIPTION
**Pull request by PointKernel**
## Description

Closes #10941 

This PR refactors the CSV reader benchmarks with nvbench and reduces the number of test cases by isolating data type, IO type, column selection, and row selection.
 
Example output of the new benchmarks:
<details>
  <summary>Benchmark results</summary>
### [0] Quadro RTX 8000

| data_type | Samples |  CPU Time  | Noise |  GPU Time  | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|-----------|---------|------------|-------|------------|-------|------------------|-------------------|-------------------|
|  INTEGRAL |      5x |    1.140 s | 0.09% |    1.140 s | 0.09% |        235553841 |         1.202 GiB |       668.564 MiB |
|     FLOAT |      5x |    1.262 s | 0.04% |    1.262 s | 0.04% |        212718321 |         1.041 GiB |       713.885 MiB |
|   DECIMAL |      5x | 272.787 ms | 0.03% | 272.784 ms | 0.03% |        984060406 |       396.279 MiB |       167.951 MiB |
| TIMESTAMP |      7x |    1.681 s | 0.47% |    1.681 s | 0.47% |        159723724 |         2.281 GiB |       814.268 MiB |
|  DURATION |      7x |    2.121 s | 0.50% |    2.121 s | 0.50% |        126587514 |         2.588 GiB |       971.320 MiB |
|    STRING |     19x | 496.713 ms | 0.50% | 496.710 ms | 0.50% |        540426462 |       859.526 MiB |       277.082 MiB |

### [0] Quadro RTX 8000

|     io      | Samples | CPU Time | Noise | GPU Time | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|-------------|---------|----------|-------|----------|-------|------------------|-------------------|-------------------|
|    FILEPATH |      9x |  1.185 s | 0.49% |  1.185 s | 0.49% |        226466264 |         1.445 GiB |       618.876 MiB |
| HOST_BUFFER |      5x |  1.170 s | 0.14% |  1.170 s | 0.14% |        229459856 |         1.445 GiB |       618.876 MiB |

### [0] Quadro RTX 8000

| column_selection | row_selection | Samples | CPU Time | Noise | GPU Time | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|------------------|---------------|---------|----------|-------|----------|-------|------------------|-------------------|-------------------|
|              ALL |           ALL |      5x |  1.246 s | 0.18% |  1.246 s | 0.18% |        215514992 |         1.582 GiB |       653.520 MiB |
|        ALTERNATE |           ALL |      5x |  1.128 s | 0.08% |  1.128 s | 0.08% |        119009844 |         1.116 GiB |       648.908 MiB |
|       FIRST_HALF |           ALL |      5x |  1.143 s | 0.07% |  1.143 s | 0.07% |        117443933 |         1.121 GiB |       653.520 MiB |
|      SECOND_HALF |           ALL |      5x |  1.152 s | 0.16% |  1.152 s | 0.16% |        116478469 |         1.121 GiB |       653.520 MiB |

### [0] Quadro RTX 8000

| column_selection | row_selection | num_chunks | Samples | CPU Time | Noise | GPU Time | Noise | bytes_per_second | peak_memory_usage | encoded_file_size |
|------------------|---------------|------------|---------|----------|-------|----------|-------|------------------|-------------------|-------------------|
|              ALL |    BYTE_RANGE |          1 |      5x |  1.244 s | 0.16% |  1.244 s | 0.16% |        215763257 |         1.582 GiB |       653.520 MiB |
|              ALL |    BYTE_RANGE |          8 |      5x |  1.170 s | 0.04% |  1.170 s | 0.04% |        229339594 |       202.596 MiB |       653.520 MiB |
|              ALL |         NROWS |          1 |      5x |  1.244 s | 0.12% |  1.244 s | 0.12% |        215808401 |         1.582 GiB |       653.520 MiB |
|              ALL |         NROWS |          8 |      4x |  4.560 s |  inf% |  4.560 s |  inf% |         58870122 |       320.771 MiB |       653.520 MiB |
|              ALL |    SKIPFOOTER |          1 |      5x |  1.245 s | 0.10% |  1.245 s | 0.10% |        215660012 |         1.582 GiB |       653.520 MiB |
|              ALL |    SKIPFOOTER |          8 |      3x |  7.443 s |  inf% |  7.443 s |  inf% |         36065528 |         1.269 GiB |       653.520 MiB |

</details>

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/benchmarks/io/csv/csv_reader_input.cpp [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+  cudf::io::write_csv(options);
+
+  cudf::io::csv_reader_options const read_options =
+    cudf::io::csv_reader_options::builder(source_sink.make_source_info());
+
+  auto mem_stats_logger = cudf::memory_stats_logger();  // init stats logger
+  state.set_cuda_stream(nvbench::make_cuda_stream_view(cudf::default_stream_value.value()));
+  state.exec(nvbench::exec_tag::sync | nvbench::exec_tag::timer,
+             [&](nvbench::launch& launch, auto& timer) {
+               try_drop_l3_cache();
```
Comment (ttnghia) [inline cpp/benchmarks/io/csv/csv_reader_input.cpp:53]: Maybe need comments on why is this?

File: cpp/benchmarks/io/csv/csv_reader_options.cpp [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+constexpr size_t data_size = 256 << 20;
+
+template <column_selection ColSelection, row_selection RowSelection>
+void BM_csv_read_varying_options(
+  nvbench::state& state,
+  nvbench::type_list<nvbench::enum_type<ColSelection>, nvbench::enum_type<RowSelection>>)
+{
+  cudf::rmm_pool_raii rmm_pool;
+
+  auto constexpr num_chunks = 1;
```
Comment (ttnghia) [inline cpp/benchmarks/io/csv/csv_reader_options.cpp:37]: Comments on why is this?

Comment (PointKernel) [reply cpp/benchmarks/io/csv/csv_reader_options.cpp:37]: @vuule Do we want benchmarks varying `num_chunks` in this case?

Comment (vuule) [reply cpp/benchmarks/io/csv/csv_reader_options.cpp:37]: We should use different numbers of chunks in the "csv_read_row_selection" benchmark. The old code used 1 and 8, we keep it that way for now.

**Review by ttnghia — APPROVED** [review]

File: cpp/benchmarks/io/csv/csv_reader_options.cpp (hunk 2 above)
Comment (PointKernel) [reply cpp/benchmarks/io/csv/csv_reader_options.cpp]: Done

**Review by vuule — APPROVED** [review]
:fire:

**Comment by PointKernel** [comment]
@gpucibot merge

**Pull request merged**